content/blog/insecure-software-development/index.md

```diff
@@ -0,0 +1,120 @@
+have in common? I will blame _hidden complexity_ here.
+
+Log4J abstracted environment variables, JNDI, and other kinds of lookups in log messages through a custom syntax that is parsed at the time of logging. Dependency confusion was caused by package managers trying to abstract the domain of dependency management, which is far from simple. In both cases library maintainers made the behaviour implicit.
+
+## Complexity
+
+In practical terms we can define complexity as anything pertaining to our system that makes it difficult to understand or modify.
+
+As software evolves and gains more features, it becomes complex, with subtle coupling between components. This leads to bugs and longer development cycles. Eventually no single developer can keep the entire context of their codebases in their heads anymore and more bugs start popping up, slowing the development cycle even more. There are usually two approaches to fight complexity:
+
+1. Make the code simpler and obvious
+2. Encapsulate complexity, abstract
+
+### Make the code simpler and obvious
+
+The idea here is to make things _explicit_. The less our users have to guess in order to use what we built the better. Users here can range from developers to end users of our products, libraries or our own codebases.
+
+1. Good naming practices, meaningful names
+   - Can we understand on a high level what this code is trying to achieve without having to dig deeper?
+2. Eliminating special cases
+   - Can we enforce invariants?
+   - Can we just crash instead of handling special cases?
+   - Can we have a single implementation for our interfaces?
+3. Ensuring consistency
+   - Are we following the same patterns across the codebase?
+   - Are we enforcing our conventions?
+   - Are our comments high level and useful?
+   - Do we have good defaults?
+
+From a security point of view fighting complexity by making the code simpler and obvious is a clear winner. Why should we encapsulate parsing environment variables from log messages when this is clearly a special case? Developers can explicitly handle this complexity instead of relying on the library.
+
+### Encapsulate complexity, abstract
+
+The idea of this approach is to _hide complexity_ so developers don't need to understand all the complexities of a certain domain in order to work on it. This is usually related to classes or modules, where each one is designed to be relatively independent of each other. Developers can work in a module without concerning themselves with another.
+
+Notice that this goes against our main point of making things explict. There is a fine balance between choosing what should be explicit and which parts of the domain we can hide without compromising our decision making process while coding. Let's investigate how things can go wrong.
+
+Abstractions can go wrong in two ways:
```

Comment:
Its nice that we can see things that would make an abstraction poor, but it'd be great to note characteristics for make for a good abstraction. Is it limited to just the negation to the points below? or are there other important traits